### ISSUE TITLE
Arrow keys scroll the page with SDL2

### ISSUE THREAD

**Issue by iboB** [emscripten-core/emscripten#3665]
As specified in the docs the canvas is supposed to take the focus (unless I specify `SDL_EMSCRIPTEN_KEYBOARD_ELEMENT`). While it does correctly capture backspace for example, the arrow keys are still sent to the browser, which causes scrolling of the page when the intention is to send commands to the js application.

Is there a way to stop sending the arrow key events to the browser?

**Comment by kripken** [comment]
Do you have a testcase showing the issue? It works when I try e.g. `tests/sdl2_text.c`, the page doesn't scroll with the arrow keys.

**Comment by iboB** [comment]
Here is a small repro program: https://github.com/iboB/ibob.github.com/blob/master/tests/emscripten/sdl.cpp

You can see it here running too: `ibob.github.io/tests/emscripten/sdl.html`
This is compiled in Windows with the line: `> emcc sdl.cpp -o sdl.html -O2 -std=c++11 -s USE_SDL=2`

When I open the html in Google Chrome or MS IE 11 in a small-enough window, pressing the arrow keys will also scroll the page

**Comment by iboB** [comment]
This cannot be reproduced in Firefox

**Comment by kripken** [comment]
I can confirm, I see scrolling in chrome but not in firefox.

No idea which browser is right here. We should probably make a tiny (non-compiled) testcase and file issues on the browser vendors.

**Comment by kripken** [comment]
I filed https://github.com/emscripten-ports/SDL2/issues/21 , let's continue the discussion there as this is an SDL2 issue.

**Comment by rtrussell** [comment]
> Unless we want to open an upstream SDL issue

I can't see anything obviously wrong with the code in [SDL_emscriptenevents.c](https://github.com/libsdl-org/SDL/blob/main/src/video/emscripten/SDL_emscriptenevents.c), which still seems to check the navigation keys:

```
        SDL_bool prevent_default = SDL_TRUE;
        ....
        if ((scancode == SDL_SCANCODE_BACKSPACE) ||
        (scancode == SDL_SCANCODE_TAB) ||
        (scancode == SDL_SCANCODE_LEFT) ||
        (scancode == SDL_SCANCODE_UP) ||
        (scancode == SDL_SCANCODE_RIGHT) ||
        (scancode == SDL_SCANCODE_DOWN) ||
        ((scancode >= SDL_SCANCODE_F1) && (scancode <= SDL_SCANCODE_F15)) ||
        keyEvent->ctrlKey) {
        is_nav_key = SDL_TRUE;
    }

    if ((eventType == EMSCRIPTEN_EVENT_KEYDOWN) && SDL_EventEnabled(SDL_EVENT_TEXT_INPUT) && !is_nav_key) {
        prevent_default = SDL_FALSE;
    }

    return prevent_default;
```
What is seemingly supposed to happen is that if the callback function passed to `emscripten_set_keydown_callback()` returns **true** the native (canvas) action of the key is suppressed.  It appears that the above code **does** return true for the navigation keys.

**Comment by sbc100** [comment]
Does the suggested fix looks right to you: https://github.com/emscripten-ports/SDL2/issues/21#issuecomment-144998911

We have a feature where you can try `EMCC_LOCAL_PORTS="sdl2=/home/alon/Dev/ports/SDL2"` to try out your own fork of SDL2 if you want to try it out.    Otherwise mabye @Daft-Freak can pick it up again?

**Comment by rtrussell** [comment]
> Does the suggested fix looks right to you: [emscripten-ports/SDL2#21 (comment)](https://github.com/emscripten-ports/SDL2/issues/21#issuecomment-144998911)

That is the code **already** in SDL2 (listed above), isn't it?   That was my point, superficially SDL looks to be correct, so is the problem that Emscripten is not acting properly on the return value?

**Comment by sbc100** [comment]
I don't know I was just going by the last comment left by @Daft-Freak: "Looks like I never actually applied my fix for arrow keys. I'll try to look at this again soon."   Perhaps it landed but never got tested.  I let them respond.

**Comment by Daft-Freak** [comment]
I think it did get in eventually. The keyboard code had some large changes recently, but I don't think the port has been updated to a version with those yet.

(Will have to check tomorrow, it's a bit late for digging through history now)

**Comment by rtrussell** [comment]
> the last comment left by @Daft-Freak: "Looks like I never actually applied my fix for arrow keys.

The last commit by @Daft-Freak is [this](https://github.com/libsdl-org/SDL/commit/9de97e19ccd8f97989a78629fe08d47fb758c97f), as far as I can see.

**Comment by Daft-Freak** [comment]
Yep, https://github.com/libsdl-org/SDL/commit/126e5a1dbfca7df34da9be7d537ca0672857efbf and the keyboard changes I was thinking of aren't in 2.24.x. So possibly something else broke.

**Comment by rtrussell** [comment]
> the keyboard changes I was thinking of aren't in 2.24.x. So possibly something else broke.

What I'd like to establish is whether the issue is, nevertheless, in SDL2 (in which case I can probably work around it myself) or is it in Emscripten, when I probably can't.  Is there a quick way you could check whether returning **true** from [Emscripten_HandleKey](https://github.com/libsdl-org/SDL/blob/6895e1700f78735bda8cdf5156e05954f44907e0/src/video/emscripten/SDL_emscriptenevents.c#L785) does successfully suppress the local (canvas) action?

**Comment by Daft-Freak** [comment]
Quickest way to check is probably with `EMCC_LOCAL_PORTS="sdl2=/path/to/modified/SDL2`. Though I did quickly test with the version I have installed (3.1.26) + some minimal code and it seemed to be working. (No scrolling with the arrow keys)

```c
#include <emscripten.h>

#include "SDL.h"

void loop()
{
  SDL_Event event;
  while(SDL_PollEvent(&event));
}

int main(int argc, char *argv[])
{
  SDL_Init(SDL_INIT_VIDEO);

  SDL_CreateWindow("Test", SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, 640, 480, 0);

  emscripten_set_main_loop(loop, 0, 1);

  return 0;
}
```

**Comment by rtrussell** [comment]
> Quickest way to check is probably with EMCC_LOCAL_PORTS="sdl2=/path/to/modified/SDL2

So what am I supposed to set this path to, exactly?  If I set it to the root of the standard SDL2 project I just get lots of Python errors, for example an error related to 'freetype' (which is **SDL2_ttf**, not SDL2): "ValueError: not enough values to unpack (expected 2, got 1)".

**Comment by Daft-Freak** [comment]
Hmm, that should be it. Seems to be working for me.

`EMCC_LOCAL_PORTS="sdl2=/home/daftfreak/SDL2/SDL2-upstream-git" emcc -sUSE_SDL=2 ...`

(that's a semi-recent clone of upstream)

**Comment by ericoporto** [comment]
I believe this is fixed in current SDL2 in Emscripten.

**Comment by sbc100** [comment]
Lets close this out. Please reopen if you are still effected with latest emscirpten (e.g. `emsdk install tot`)
